tests/it/query_syntax.rs

```diff
@@ -0,0 +1,199 @@
+    let value = client
+        .query("--comment\nSELECT *\n--comment\n/*\nпроверка\t;;\r\n;;\n*/\nFROM (select 1)\n;")
+        .fetch_one::<u8>()
+        .await
+        .unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_semicolon() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT * FROM system.numbers LIMIT 3;")
+        .fetch_all::<u64>()
+        .await
+        .unwrap();
+    assert_eq!(value, [0, 1, 2]);
+}
+
+#[tokio::test]
+async fn query_with_semicolon_trailing() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT * FROM system.numbers LIMIT 3;;;;;;;;;;;;;;;;;")
+        .fetch_all::<u64>()
+        .await
+        .unwrap();
+    assert_eq!(value, [0, 1, 2]);
+}
+
+#[tokio::test]
+async fn query_with_semicolon_in_comment() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT * FROM system.numbers LIMIT 3 -- comment with ;;")
+        .fetch_all::<u64>()
+        .await
+        .unwrap();
+    assert_eq!(value, [0, 1, 2]);
+}
+
+#[tokio::test]
+async fn query_with_semicolon_in_multiline_comment() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT * FROM system.numbers LIMIT 3 /*\n * comment with --\n * and ; inside\n*/")
+        .fetch_all::<u64>()
+        .await
+        .unwrap();
+    assert_eq!(value, [0, 1, 2]);
+}
+
+#[tokio::test]
+async fn query_with_semicolon_allowed() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT ';;';;")
+        .fetch_one::<String>()
+        .await
+        .unwrap();
+    assert_eq!(value, ";;");
+}
+
+#[tokio::test]
+async fn query_with_tail_newline() {
+    let client = prepare_database!();
+    let value = client.query("SELECT 1\n").fetch_one::<u8>().await.unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_tail_space() {
+    let client = prepare_database!();
+    let value = client.query("SELECT 1 ").fetch_one::<u8>().await.unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_tail_tab() {
+    let client = prepare_database!();
+    let value = client.query("SELECT 1\t").fetch_one::<u8>().await.unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_tail_carriage_return() {
+    let client = prepare_database!();
+    let value = client.query("SELECT 1\r").fetch_one::<u8>().await.unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_tail_carriage_return_newline() {
+    let client = prepare_database!();
+    let value = client
+        .query("SELECT 1\r\n")
+        .fetch_one::<u8>()
+        .await
+        .unwrap();
+    assert_eq!(value, 1);
+}
+
+#[tokio::test]
+async fn query_with_tail_carriage_return_space() {
+    let client = prepare_database!();
+    let value = client.query("SELECT 1\r ").fetch_one::<u8>().await.unwrap();
+    assert_eq!(value, 1);
+}
```

Comment:
Let's add one more test: with a user that has readonly=1 in the profile. TLDR how to create one: see JS client tests https://github.com/ClickHouse/clickhouse-js/blob/main/packages/client-common/__tests__/fixtures/read_only_user.ts#L31-L43

This is to ensure that the read-only users are properly supported and that the existing code will not be broken in these cases.

Reply:
Added.